### ISSUE TITLE
uart shell: ring buffer usage bug

### ISSUE THREAD

**Issue by qianfan-Zhao** [zephyrproject-rtos/zephyr#11281]
I had test http_server example on frdm_k64f/custom SAME70 board, Zephyr board has the default ipv4 config: 192.0.2.1, the host(ubuntu) computer's ip address is 192.0.2.2.

1. After power up, ping Zephyr from Ubuntu successful.
```
➜  ~ ping 192.0.2.1 
PING 192.0.2.1 (192.0.2.1) 56(84) bytes of data.
64 bytes from 192.0.2.1: icmp_seq=1 ttl=64 time=0.282 ms
64 bytes from 192.0.2.1: icmp_seq=2 ttl=64 time=0.209 ms
```

2. Ping Ubuntu from Zephyr, failed.
```
uart:~$ net ping 192.0.2.2
Sent a pinSg to 192.0.2.2
Ping timeout
uart:~$ 
```

(Please notice the **S** in Send a ping**S**g to, I don't know where it comes)

3. Ping Zephyr from Ubuntu again, failed.

**Comment by qianfan-Zhao** [comment]
I had try revery commit: [59c23b1] 'shell: wildcard active by default' on the laster master branch(45f0917)

1. After power up, ping Zephyr from Ubuntu successful.
```
➜  ~ ping 192.0.2.1 
PING 192.0.2.1 (192.0.2.1) 56(84) bytes of data.
64 bytes from 192.0.2.1: icmp_seq=1 ttl=64 time=0.282 ms
64 bytes from 192.0.2.1: icmp_seq=2 ttl=64 time=0.209 ms
```

2. Ping Ubuntu from Zephyr, failed.
```
uart:~$ net ping 192.0.2.2
Sent a pinSg to 192.0.2.2
Ping timeout
uart:~$ 
```

3. Ping Zephyr from Ubuntu again, failed.

Use bisect to check commit again after revert commit [59c23b1]. Next is the git bisect log:

```
➜  zephyr git:(6575e84002) git bisect bad 
6575e84002f00c97c99bc947e3a395fb59d701fb is the first bad commit
commit 6575e84002f00c97c99bc947e3a395fb59d701fb
Author: Michael Scott <mike@foundries.io>
Date:   Fri Oct 19 14:14:27 2018 -0700

    logging: start log processing thread in enable_logger()
    
    Currently, if CONFIG_LOG_PROCESS_THREAD is enabled, log processing
    does not start until just before the main() function is called in
    the user sample.  This means that no real-time logging information
    will be sent to the backends while drivers and/or other processes
    which are triggered during kernel boot are processing their tasks.
    Worse, if they take a longish time to finish, the user is
    presented with a blank log backend regardless of the debug levels.
    
    Let's start processing logs earlier to avoid this issue.
    
    NOTE: Since this thread is now triggered specifically rather than
    started during static thread init, let's also name it.
    
    Signed-off-by: Michael Scott <mike@foundries.io>

:040000 040000 ec9e8507174b04716c2d12fed6bbb2340e231085 30f6b5bed7c310ee620858b1501f839d50606b68 M      subsys

git bisect start
# good: [9a69cfbf19edc8726f60dbbb36a776ce4e259992] mgmt: Add function to unregister GATT service
git bisect good 9a69cfbf19edc8726f60dbbb36a776ce4e259992
# bad: [45f09173082980d6e83950d15d344a6f18387abf] dts: intel_s1000: update SRAM size to 4MB
git bisect bad 45f09173082980d6e83950d15d344a6f18387abf
# bad: [0757b61c4a0ac099a35b7730a3069d7cc5a9eacc] boards: efr32mg_sltb004a: Add debug capability to Silabs TB2
git bisect bad 0757b61c4a0ac099a35b7730a3069d7cc5a9eacc
# bad: [4ab023d31f4f7da7af50d31557bc4e3b839a5814] doc: Fix licensing links for cmsis, nordic, and mcux components
git bisect bad 4ab023d31f4f7da7af50d31557bc4e3b839a5814
# bad: [1df011c8e2e2e7ea82955b7c477d7f84b382c1be] coccinelle: Add script to find cases of missing locks
git bisect bad 1df011c8e2e2e7ea82955b7c477d7f84b382c1be
# bad: [39c3c4e02013ff2e2e9b4afb3268562a78a00957] CODEOWNERS: Update list to add `scripts/coccinelle/` directory
git bisect bad 39c3c4e02013ff2e2e9b4afb3268562a78a00957
# bad: [4a2837f8667b06097da327d89cea2208669ca2a7] arch: arm: sam0: enable SEGGER RTT on Atmel sam0 SoCs
git bisect bad 4a2837f8667b06097da327d89cea2208669ca2a7
# bad: [6575e84002f00c97c99bc947e3a395fb59d701fb] logging: start log processing thread in enable_logger()
git bisect bad 6575e84002f00c97c99bc947e3a395fb59d701fb
# first bad commit: [6575e84002f00c97c99bc947e3a395fb59d701fb] logging: start log processing thread in enable_logger()
```

**Comment by qianfan-Zhao** [comment]
I had check the tag: v1.13.0 is a good commit, and the latest commmit(45f0917) on master is a bad commit.
Check the first commit which broke net by using 'git bisect', next is the log:

```
➜  zephyr git:(4962618e5f) git bisect bad 
4962618e5f7074be3b4fffc9c7db60b17761b6d7 is the first bad commit
commit 4962618e5f7074be3b4fffc9c7db60b17761b6d7
Author: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>
Date:   Mon Sep 10 09:48:56 2018 +0200

    shell: shell_uart: add ring_buffers and interrupt support
    
    Improved RX path to use ring buffer for incoming data instead of single
    byte buffer. Improved TX path to use ring buffer. Added support for
    asynchronous UART API (interrupts).
    
    Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

:040000 040000 3dbea7f34c5ac8b67432f24104b2e21eba7d0e31 ebe3b5de71c1e9926aca84f147e19be0f65c668e M      include
:040000 040000 b75aa6d6abe08a81036a73f70851dc2b3d2a4dc0 b6d58dddd166f53cd2a240bba6b3496bc408814f M      samples
:040000 040000 f669c82efc04bf7f9ece0b5fd6372d9f3104b1d2 d8fd705dc99358b02c4a1500046a99618f2e6add M      subsys
➜  zephyr git:(4962618e5f) git bisect log 
git bisect start
# good: [d62e861a01a2f4455ae89d105aa9dbc4bef501c5] release: Zephyr 1.13.0
git bisect good d62e861a01a2f4455ae89d105aa9dbc4bef501c5
# bad: [45f09173082980d6e83950d15d344a6f18387abf] dts: intel_s1000: update SRAM size to 4MB
git bisect bad 45f09173082980d6e83950d15d344a6f18387abf
# good: [44f0a79dbb75634480f9da299edacaefaac28473] drivers: i2c: enable logging for drivers loading i2c-priv.h
git bisect good 44f0a79dbb75634480f9da299edacaefaac28473
# good: [c0f893480aeffd7421bfcc7dbe32e701f59fb6fa] samples: mesh: nrf52: short time multireset bt-mesh unprovisioning
git bisect good c0f893480aeffd7421bfcc7dbe32e701f59fb6fa
# good: [b4caa74684a4bd4a4672f7c71256ec27af05bf84] subsys: filesystem: move conditional location
git bisect good b4caa74684a4bd4a4672f7c71256ec27af05bf84
# good: [5addb3a9250fa6a6a3ca0a93e2eb2ad8cb4f39d8] samples: net: Correct FRDM + CR20A configuration
git bisect good 5addb3a9250fa6a6a3ca0a93e2eb2ad8cb4f39d8
# bad: [602e08c69ac4caa045693d9d00ae68a64bc57186] Bluetooth: host: Don't restart scan if connection is pending
git bisect bad 602e08c69ac4caa045693d9d00ae68a64bc57186
# bad: [59c23b11ffd618cdade0f80aadea37b001b1f771] shell: wildcard active by default
git bisect bad 59c23b11ffd618cdade0f80aadea37b001b1f771
# good: [b6979010c6afe131cb62fb298789fa1466849c2c] Bluetooth: host: Fallback to L2CAP for CPUP if rejected by master on LL
git bisect good b6979010c6afe131cb62fb298789fa1466849c2c
# good: [899d704dee83afde9dd69272103adc4616aec3d8] Bluetooth: Add basic documentation for address types
git bisect good 899d704dee83afde9dd69272103adc4616aec3d8
# good: [523caef224f5857b91e12d1a301aa901dac6767e] subsys: nvs: fix ATE read/write when write block size > 8 bytes
git bisect good 523caef224f5857b91e12d1a301aa901dac6767e
# good: [79095099bbb1b6d9e8a4732b651aacd78f10266e] samples: bmg160: use default log level
git bisect good 79095099bbb1b6d9e8a4732b651aacd78f10266e
# bad: [4962618e5f7074be3b4fffc9c7db60b17761b6d7] shell: shell_uart: add ring_buffers and interrupt support
git bisect bad 4962618e5f7074be3b4fffc9c7db60b17761b6d7
# first bad commit: [4962618e5f7074be3b4fffc9c7db60b17761b6d7] shell: shell_uart: add ring_buffers and interrupt support
➜  zephyr git:(4962618e5f) 


uart:~*$* *[*m Booting Zephyr OS zephyr-v1.13.0-1667-g4962618e5f *****

uart:~$ net ping 192.0.2.2
Sent a ping Sto 192.0.2.2
Ping timeout
```

**Comment by qianfan-Zhao** [comment]
I had disabled CONFIG_UART_INTERRUPT_DRIVEN, the 'net ping' worked again.

**Comment by qianfan-Zhao** [comment]
The usage of ring_buffer in funtion uart_rx_handle is wrong.
ring_buf_put_claim will return 0 if the tail is equal ring buffer's size.

```
static void uart_rx_handle(const struct shell_uart *sh_uart)
{
	u8_t *data;
	u32_t len;
	u32_t rd_len;
	bool new_data = false;

	do {
		len = ring_buf_put_claim(sh_uart->rx_ringbuf, &data,
					 sh_uart->rx_ringbuf->size);
		rd_len = uart_fifo_read(sh_uart->ctrl_blk->dev, data, len);

		if (rd_len) {
			new_data = true;
		}

		ring_buf_put_finish(sh_uart->rx_ringbuf, rd_len);
	} while (rd_len && (rd_len == len));

	if (new_data) {
		sh_uart->ctrl_blk->handler(SHELL_TRANSPORT_EVT_RX_RDY,
					   sh_uart->ctrl_blk->context);
	}
}
```

For most SoC's(without hardware FIFO) drivers implements, they always save the
received uart byte to ring buffer and return 1, doesn't check the 'size' param
in uart_fifo_read even if it is zero.

Such as the implements in usart_sam.c:
```
static int usart_sam_fifo_fill(struct device *dev, const uint8_t *tx_data,
			       int size)
{
	volatile Usart * const usart = DEV_CFG(dev)->regs;

	/* Wait for transmitter to be ready. */
	while ((usart->US_CSR & US_CSR_TXRDY) == 0) {
		;
	}

	usart->US_THR = *tx_data;

	return 1;
}
```

The save action in usart_sam_fifo_fill will cause buffer overflow, due to
the buffer pointer is a illege pointer. Actually the write action will write
shell_uart_rx_ringbuf[SHELL_UART_RX_BUF_SIZE], that maybe changed another
variable.

**Comment by nordic-krch** [comment]
@qianfan-Zhao 

It may return 0 only if ringbuffer is full. That case is in fact not handled here. Overrun handling must be added here.

regarding:
> ring_buf_put_claim will return 0 if the tail is equal ring buffer's size.

If this is true then it's a bug in ring buffer. In that case it should wrap and return pointer to the start of the buffer. I will check it.

**Comment by qianfan-Zhao** [comment]
> If this is true then it's a bug in ring buffer. In that case it should wrap and return pointer to the start of the buffer. I will check it.

Unfortunately the ring buffer will not wrap automatilly, it will wrap when ring_buf_put_finish.
Please check the ring buffer APIs

**Comment by qianfan-Zhao** [comment]
@nordic-krch 
> If this is true then it's a bug in ring buffer. 

I think it is not a bug. It is such a design.
ring_buf_put_claim want provide a memory buffer continuous address, to make sure the DMA can works.

**Comment by tbursztyka** [comment]
if it's by design then it's not documented, like at all.

**Comment by nordic-krch** [comment]
@qianfan-Zhao , are you using ring buf size = 1?

**Comment by qianfan-Zhao** [comment]
@nordic-krch No I used the default value in Kconfig.

**Comment by pfalcon** [comment]
So, this unlikely should be assigned to me, as the report is against the shell behavior, so it rather goes @jarz-nordic/@nordic-krch way. Of course, I can host it for awhile, it's on my todo to compare subsys/console and shell's UART ISR implementation. I'm just not sure when I get to it...

**Comment by tbursztyka** [comment]
@pfalcon reassign it, it's definitely shell/ring buffer which is the issue here.

**Comment by pfalcon** [comment]
> @pfalcon reassign it, it's definitely shell/ring buffer which is the issue here.

Indeed, goes to @nordic-krch .

**Comment by nashif** [comment]
@nordic-krch what is the status of this one?

**Comment by nordic-krch** [comment]
@qianfan-Zhao could you retest it? There was a bug in UART driver that has been fix #11778.

**Comment by nordic-krch** [comment]
I'm closing this one since there is no feedback from @qianfan-Zhao and it's likely that it has been fixed by #11778
